source/opt/local_single_store_elim_pass.cpp

```diff
@@ -0,0 +1,445 @@
+        ssa_var2store_.clear();
+        return;
+      } break;
+      default:
+        break;
+      } // switch
+    }
+  }
+}
+
+void LocalSingleStoreElimPass::ReplaceAndDeleteLoad(
+    ir::Instruction* loadInst, uint32_t replId) {
+  (void) def_use_mgr_->ReplaceAllUsesWith(loadInst->result_id(), replId);
+  DCEInst(loadInst);
+}
+
+LocalSingleStoreElimPass::GetBlocksFunction
+LocalSingleStoreElimPass::AugmentedCFGSuccessorsFunction() const {
+  return [this](const ir::BasicBlock* block) {
+    auto asmi = augmented_successors_map_.find(block);
+    if (asmi != augmented_successors_map_.end())
+      return &(*asmi).second;
+    auto smi = successors_map_.find(block);
+    return &(*smi).second;
+  };
+}
+
+LocalSingleStoreElimPass::GetBlocksFunction
+LocalSingleStoreElimPass::AugmentedCFGPredecessorsFunction() const {
+  return [this](const ir::BasicBlock* block) {
+    auto apmi = augmented_predecessors_map_.find(block);
+    if (apmi != augmented_predecessors_map_.end())
+      return &(*apmi).second;
+    auto pmi = predecessors_map_.find(block);
+    return &(*pmi).second;
+  };
+}
+
+void LocalSingleStoreElimPass::CalculateImmediateDominators(
+    ir::Function* func) {
+  // Compute CFG
+  vector<ir::BasicBlock*> ordered_blocks;
+  predecessors_map_.clear();
+  successors_map_.clear();
+  for (auto& blk : *func) {
+    ordered_blocks.push_back(&blk);
+    blk.ForEachSuccessorLabel([&blk, &ordered_blocks, this](uint32_t sbid) {
+      successors_map_[&blk].push_back(label2block_[sbid]);
+      predecessors_map_[label2block_[sbid]].push_back(&blk);
+    });
+  }
+  // Compute Augmented CFG
+  augmented_successors_map_.clear();
+  augmented_predecessors_map_.clear();
+  successors_map_[&pseudo_exit_block_] = {};
+  predecessors_map_[&pseudo_entry_block_] = {};
+  auto succ_func = [this](const ir::BasicBlock* b)
+    { return &successors_map_[b]; };
+  auto pred_func = [this](const ir::BasicBlock* b)
+    { return &predecessors_map_[b]; };
+  CFA<ir::BasicBlock>::ComputeAugmentedCFG(
+    ordered_blocks,
+    &pseudo_entry_block_,
+    &pseudo_exit_block_,
+    &augmented_successors_map_,
+    &augmented_predecessors_map_,
+    succ_func,
+    pred_func);
+  // Compute Dominators
+  vector<const ir::BasicBlock*> postorder;
+  auto ignore_block = [](cbb_ptr) {};
+  auto ignore_edge = [](cbb_ptr, cbb_ptr) {};
+  spvtools::CFA<ir::BasicBlock>::DepthFirstTraversal(
+    ordered_blocks[0], AugmentedCFGSuccessorsFunction(),
+    ignore_block, [&](cbb_ptr b) { postorder.push_back(b); },
+    ignore_edge);
+  auto edges = spvtools::CFA<ir::BasicBlock>::CalculateDominators(
+    postorder, AugmentedCFGPredecessorsFunction());
+  idom_.clear();
+  for (auto edge : edges)
+    idom_[edge.first] = edge.second;
+}
+
+bool LocalSingleStoreElimPass::Dominates(
+    ir::BasicBlock* blk0, uint32_t idx0,
+    ir::BasicBlock* blk1, uint32_t idx1) {
+  if (blk0 == blk1)
+    return idx0 < idx1;
```

Comment:
An instruction always dominates itself.  You probably don't care about that case, but I'd like to be consistent with the standard definition.